File: tests/jnet/TestSecureNetworking.java
import java.util.Date;
import java.util.ArrayList;
import java.io.IOException;
import org.junit.Test;
import org.junit.Assert;
import org.junit.Assume;
import org.junit.Before;
import org.junit.After;
import jnet.Bytes;
import jnet.secure.JSSLServer;
import jnet.secure.JSSLClientSocket;
import jnet.JClientSocket;


public class TestSecureNetworking {

    private class EchoServer extends JSSLServer {
        public EchoServer(String ip, int port) throws IOException {
            super(ip, port);
        }

        @Override
        public void clientConnected(JClientSocket clientSocket) {}

        @Override
        public void clientDisconnected(JClientSocket clientSocket) {}

        @Override
        public void clientCommunicated(byte[] recv, JClientSocket clientSocket) {
            try {
                this.send(recv, clientSocket);
            }
            catch (IOException e) {
                Assert.fail("send in EchoServer threw IOException: " + e);
            }
        }
    }


    @Test
    public void testServerConnectDisconnect() throws IOException {
        JSSLServer server = new EchoServer("localhost", 9000);
        server.close();
    }

    @Test
    public void testClientConnectDisconnect() throws IOException {
        JSSLServer server = new EchoServer("localhost", 9000);
        JSSLClientSocket client = new JSSLClientSocket();
        client.connect(server.getIP(), server.getPort());
        client.close();
        server.close();
    }

    @Test
    public void testClientSendRecvBytes() throws IOException {
        JSSLServer server = new EchoServer("localhost", 9000);
        JSSLClientSocket client = new JSSLClientSocket();
        client.connect(server.getIP(), server.getPort());

        for (int i = 0; i < TestMain.RANDOM_TEST_AMOUNT; i++) {
            byte[] expected = TestMain.randomBytes(TestMain.RANDOM_DIS.nextInt(1024));
            client.send(expected);
            byte[] generated = client.recv();
            Assert.assertArrayEquals(generated, expected);
        }

        client.close();
        server.close();
    }

    @Test
    public void testClientSendRecvString() throws IOException {
        JSSLServer server = new EchoServer("localhost", 9000);
        JSSLClientSocket client = new JSSLClientSocket();
        client.connect(server.getIP(), server.getPort());

        for (int i = 0; i < TestMain.RANDOM_TEST_AMOUNT; i++) {
            String expected = TestMain.randomString(TestMain.RANDOM_DIS.nextInt(1024));
            client.send(expected);
            String generated = client.srecv();
            Assert.assertEquals(generated, expected);
        }

        client.close();
        server.close();
    }

}

File: tests/pnet/test_secure_networking.py
import os
import random
import pytest
import conftest
import socket
from pnet import byteutils
from pnet.secure.psslserver import PSSLServer
from pnet.secure.psslclientsocket import PSSLClientSocket


class EchoServer(PSSLServer):
    def client_connected(self, client_socket: PSSLClientSocket):
        pass

    def client_disconnected(self, client_socket: PSSLClientSocket):
        pass

    def client_communicated(self, recv: bytes, client_socket: PSSLClientSocket):
        self.send(recv, client_socket)


certfile: str = os.path.abspath(os.environ["PNET_CERTFILE"])
keyfile: str = os.path.abspath(os.environ["PNET_KEYFILE"])


class TestNetworking:

    def test_server_connect_disconnect(self):
        server: PSSLServer = EchoServer("localhost", 9000, certfile, keyfile)
        server.close()

    def test_client_connect_disconnect(self):
        server: PSSLServer = EchoServer("localhost", 9000, certfile, keyfile)
        client: PSSLClientSocket = PSSLClientSocket(certfile)
        client.connect(server.get_ip(), server.get_port())
        client.close()
        server.close()

    def test_client_send_recv_bytes(self):
        server: PSSLServer = EchoServer("localhost", 9000, certfile, keyfile)
        client: PSSLClientSocket = PSSLClientSocket(certfile)
        client.connect(server.get_ip(), server.get_port())

        for i in range(conftest.RANDOM_TEST_AMOUNT):
            expected: bytes = conftest.random_bytes(random.randint(0, 1024))
            client.send(expected)
            generated: bytes = client.recv()
            assert generated == expected

        client.close()
        server.close()

    def test_client_send_recv_str(self):
        server: PSSLServer = EchoServer("localhost", 9000, certfile, keyfile)
        client: PSSLClientSocket = PSSLClientSocket(certfile)
        client.connect(server.get_ip(), server.get_port())

        for i in range(conftest.RANDOM_TEST_AMOUNT):
            expected: str = conftest.random_str(random.randint(0, 1024))
            client.send(expected)
            generated: str = client.srecv()
            assert generated == expected

        client.close()
        server.close()
